File: articles/statistics-uncertainty-measurement-biology/python/measurement_error_simulation.py
"""
Measurement error simulation workflow.

Run:
    python python/measurement_error_simulation.py
"""

import numpy as np
import pandas as pd

from statistics_measurement_core import root_mean_squared_error


def main() -> None:
    rng = np.random.default_rng(42)

    n_samples = 200
    true_values = rng.normal(loc=10.0, scale=1.5, size=n_samples)

    systematic_bias = 0.35
    random_error_sd = 0.45

    measured_values = true_values + systematic_bias + rng.normal(
        loc=0.0,
        scale=random_error_sd,
        size=n_samples,
    )

    errors = measured_values - true_values

    summary = pd.DataFrame(
        {
            "true_mean": [true_values.mean()],
            "measured_mean": [measured_values.mean()],
            "mean_error": [errors.mean()],
            "error_sd": [errors.std(ddof=1)],
            "rmse": [root_mean_squared_error(errors)],
        }
    )

    print(summary.round(5).to_string(index=False))


if __name__ == "__main__":
    main()

File: articles/statistics-uncertainty-measurement-biology/r/measurement_error_simulation.R
# Measurement error simulation in R.

set.seed(42)

n_samples <- 200

true_values <- rnorm(n_samples, mean = 10.0, sd = 1.5)
systematic_bias <- 0.35
random_error_sd <- 0.45

measured_values <- true_values + systematic_bias + rnorm(
  n_samples,
  mean = 0,
  sd = random_error_sd
)

errors <- measured_values - true_values

summary_df <- data.frame(
  true_mean = mean(true_values),
  measured_mean = mean(measured_values),
  mean_error = mean(errors),
  error_sd = sd(errors),
  rmse = sqrt(mean(errors^2))
)

print(round(summary_df, 5))
